Add option to enable unified pipeline

graudtv opened the pull request:
Corresponding to issue #463

Option can be enabled with -u or --unified-pipeline
Works both for riscv and mips

graudtv committed:
fix codestyle

graudtv committed:
implement unified pipeline

pavelkryukov commented on simulator/modules/core/perf_sim.cpp:
```diff
@@ -13,6 +13,7 @@
 namespace config {
     static const AliasedValue<std::string> units_to_log = { "l", "logs", "nothing", "print logs for modules"};
     static const Switch topology_dump = { "tdump", "module topology dump into topology.json" };
+    const AliasedSwitch unified_pipeline = { "u", "unified-pipeline", "enable unified pipeline"};
```
Could you please localize this switch to `Execute` module?

pavelkryukov commented on simulator/modules/mem/mem.cpp:
```diff
@@ -42,8 +43,18 @@ void Mem<FuncInstr>::clock( Cycle cycle)
+        /* if unified_pipeline is enabled, there can be instructions,
+         * which do nothing on memory stage. Only dumping them */
+        assert( config::unified_pipeline);
+        sout << instr << " { unified pipeline }\n";
```
I suppose it should be dumped for all instructions, regardless of type.

graudtv committed:
add memory stage dump

graudtv committed:
move unified pipeline switch to execute.cpp

pavelkryukov commented:
Can this option be implemented without routing non-memory instructions to Memory pipeline stage?

graudtv commented:
It probably can, but it won't be very good, I suppose. In this case we won't be able to dump non-memory instructions on memory stage, so they will just disappear for one cycle after Execute, and when appear again on WriteBack
Also current solution seems to corresponds to real unified pipeline better. All instruction are passed through each stage, even if no work on some stages is done.

pavelkryukov commented:
> In this case we won't be able to dump non-memory instructions on memory stage, so they will just disappear for one cycle after Execute, and when appear again on WriteBack

That's exactly how multi-cycle instructions are modeled. They spend few cycles on Execute stage.
My question is: why do we use Memory stage for arithmetic instructions, but not the Branch stage?

> Also current solution seems to corresponds to real unified pipeline better. 

Our goal is not only to reflect real pipeline, but make implement abstract to invent non-real pipelines.

The problem with your implementation is that it makes two control points: one is steering between Memory/Writeback, and another one is scheduling.

By the way, what is the behavior if complex ALU latency is increased simultaneously with enabling Memory stage?

graudtv commented:
Could you please give an example of multi-cycle instruction?

pavelkryukov commented:
Now they are defined as:

https://github.com/MIPT-ILab/mipt-mips/blob/6e72b5383b8c06101874e6626990be92794cee9f/simulator/modules/core/perf_instr.h#L53

https://github.com/MIPT-ILab/mipt-mips/blob/bad329c173966e103b5b644ff01e92931208cd46/simulator/func_sim/operation.h#L114

For MIPS, you may check the code in unit tests (grep `is_divmult`)

https://github.com/MIPT-ILab/mipt-mips/blob/6e72b5383b8c06101874e6626990be92794cee9f/simulator/mips/t/mips32_divmult_test.cpp

graudtv commented:
> That's exactly how multi-cycle instructions are modeled

I actually a bit don't understand why it is done so. If instruction spends n cycles on execute stage, why we print "bubble" instead of something like ```execute cycle ... : ... { stage k / n }```

> Why do we use Memory stage for arithmetic instructions, but not the Branch stage?

For some reason in logs memory stage has greater priority, than branch one: memory stage is always printed, branch - only when it is executing something. It still seems a bit strange to me that arithmetic instruction between Memory and Writeback stages is "nowhere" for one cycle, so I related it to Memory.

As I understand, you suggest adding a port from Execute to Writeback with 2-cycle latency for these instructions?

> By the way, what is the behavior if complex ALU latency is increased simultaneously with enabling Memory stage?

I've checked, I've forgotten this case, will consider it

pavelkryukov commented:
> Execute to Writeback with 2-cycle latency for these instructions?

I'd configure latency of existing ports.

>  If instruction spends n cycles on execute stage, why we print "bubble" instead of something like `execute cycle ... : ... { stage k / n }`

Because such print was not implemented.

> memory stage has greater priority, than branch one: memory stage is always printed, 

I would stick to Branch implementation, since it is less verbose.

graudtv commented:
So this is ok? (writeback.cpp)
```c++
rp_execute_datapath = make_read_port<Instr>("EXECUTE_2_WRITEBACK", ( unified_pipeline) ? Port::LATENCY + 1_lt : Port::LATENCY);
```

pavelkryukov commented:
Yes, but let's use literals:

```c++
unified_pipeline ? 2_Lt : 1_Lt
```

graudtv committed:
change unified pipeline impl

graudtv commented:
I'll delete some unneeded code now

graudtv committed:
delete unneeded parts

pavelkryukov commented:
Have you checked whether bypasses and stalls operate correctly?

graudtv commented:
There is a problem with bypass scoreboard.
It takes one cycle for instructions like add to move from EXE to WB state in scoreboard, but if unified pipeline is enabled, instruction should stay there for 2 cycles (i.e. either it should wait on WB for one more cycle, or some intermediate state in scoreboard between EXE and WB should be introduced for unified pipeline). Currently register comes out of scoreboard too soon, so there are some situations when bybass works incorrectly (I can give example, if needed)

pavelkryukov commented:
> some intermediate state in scoreboard between EXE and WB should be introduced for unified pipeline)

Could you please introduce such state?

graudtv committed:
implement unified_pipeline_buffer_stage in scoreboard

Unified pipeline buffer stage in bypass scoreboard is added in order to correctly
treat bypasses and stalls when unified pipeline option is enabled.
Extra port is added, which bypasses data from
instruction between Execute and Writeback stage (from "Buffer stage") in unified pipeline to Execute stage

graudtv commented:
Now everything seems to work properly.
Some questions:
1) I didn't change the location of unified_pipeline switch, but writeback.cpp is probably not the best place for it, since the switch is used in many files now. There should it be?
2) I used the following code for debug, but it still may be useful in dumps. Should I leave it or not?
https://github.com/MIPT-ILab/mipt-mips/blob/77e37753dba032a232401cc92319089890ea210c/simulator/modules/execute/execute.cpp#L146

pavelkryukov commented:
Sorry for late response. 

> I didn't change the location of unified_pipeline switch, but writeback.cpp is probably not the best place for it, since the switch is used in many files now. There should it be?

I think it is related to Core module, and should it be in that file.

> I used the following code for debug, but it still may be useful in dumps. Should I leave it or not?

No, let's leave debug prints out of repository. They can be restored faster once debug is needed.

pavelkryukov commented:
Could you please tell whether you had a chance to run PipelineVis tool?
It would be great to see a demo of old and new modes with it, for example, for Fibonacci trace.

graudtv commented:
Demo program:
```
_start:
    add t1, x0, 0x123
    add t1, x0, 0x456
    add t2, x0, t1
    add t3, x0, t1
    add t4, t4, t4
```

Non-unified pipeline:
<img width="1346" alt="nonunified" src="https://user-images.githubusercontent.com/55806873/111652851-76e14700-8818-11eb-9cc0-274514ea54de.png">

Unified pipeline:
<img width="1334" alt="unified" src="https://user-images.githubusercontent.com/55806873/111652874-7ba5fb00-8818-11eb-86fb-2863021ba834.png">

I couldn't run Fibonacci, it fails with "Performance simulator and functional simulator executed different instructions", json file is empty

graudtv commented:
> I think it is related to Core module, and should it be in that file

I can't put it there, because core module (perf_sim.h) already includes all the other module headers, so perf_sim.h itself cannot be included inside modules to make declaration of unified_pipeline variable visible there

pavelkryukov commented:
> I couldn't run Fibonacci, it fails with "Performance simulator and functional simulator executed different instructions", 

Works for me on clean version. I used the following command line:

```./mipt-mips -b ../tests/mips/mips-fib.bin -n 10000```

Could you please confirm you use the same command line? If yes, there might be a bug in your implementation.

graudtv commented:
I had incorrect architecture option, your command line works fine
Non-unified:
<img width="1016" alt="non-unified" src="https://user-images.githubusercontent.com/55806873/112061826-1d5a7e80-8b70-11eb-93f0-7e0ca1429524.png">
Unified:
<img width="1077" alt="unified" src="https://user-images.githubusercontent.com/55806873/112061858-24818c80-8b70-11eb-882b-25b3631b1e7b.png">

pavelkryukov commented:
Could you please share IPC for both cases?

pavelkryukov commented:
Also, your diagrams show warm-up state with "cold" branch prediction. Steady state diagrams would be more informative.

graudtv commented:
> Could you please share IPC for both cases?

Non-unified:
<img width="473" alt="non-unifief" src="https://user-images.githubusercontent.com/55806873/112144089-cb563f00-8be9-11eb-9c38-d086129d4d69.png">
Unified:
<img width="452" alt="unified" src="https://user-images.githubusercontent.com/55806873/112144099-ce512f80-8be9-11eb-810e-fe358e2e0bf6.png">


> Also, your diagrams show warm-up state with "cold" branch prediction. Steady state diagrams would be more informative.

I don't see any changes in stage patterns on bigger cycles.
By the way, I can share json with you, if you would like to

pavelkryukov commented:
> I don't see any changes in stage patterns on bigger cycles.

So you see there are nops following branches, is it correct?
Why do they appear? We should predict branch direction and target perfectly, right?

> By the way, I can share json with you, if you would like to

That's your study, not mine...

graudtv commented:
> So you see there are nops following branches, is it correct?

Nops disappear after some time, you were right.
Nevertheless, there's something wrong with visualiser

-l cpu dump:
<img width="992" alt="Снимок экрана 2021-03-23 в 16 15 27" src="https://user-images.githubusercontent.com/55806873/112152826-c0081100-8bf3-11eb-8a13-3e6f1c654776.png">

-what visualiser displays:
<img width="632" alt="Снимок экрана 2021-03-23 в 16 35 57" src="https://user-images.githubusercontent.com/55806873/112155470-635a2580-8bf6-11eb-84c7-b2066540c2f9.png">

There is also no two same sequential "or"-s in -l cpu, visualiser draws two. I understand, why it happens in visualiser, but it doesn't seem very convinient

Also, there is actually no "or"-s in fibonacci program at all, there is "move" instead. I don't know what we are simulating, but it's not fibonacci for sure
objdump proof:
<img width="534" alt="Снимок экрана 2021-03-23 в 16 27 33" src="https://user-images.githubusercontent.com/55806873/112154055-f5f9c500-8bf4-11eb-981c-d5237fe05e88.png">

pavelkryukov commented:
> I don't know what we are simulating, but it's not fibonacci for sure

Again, you jumped to a quick conclusion which appears to be wrong.
This _is_ a Fibonacci sequence:

```c
int t1 = 1;
int t2 = 1;
while (true) {
   t3 = t1 + t2; // sum of two previous values
   t1 = t2; // t2 and t3 should be added on the next iteration
   t2 = t3;
}
```

> there is actually no "or"-s in fibonacci program at all, there is "move" instead.

`move $x, $y` is a pseudo-instruction, equivalent to `or $x, $y, $zero` in MIPS or RISC-V.

> -what visualiser displays:

Instruction {38} is stalled on decoder, so it may printed both on cycle 78 and 79. PipelineVis draws a bubble on cycle 78, while log prints both cycles. I agree it is not consistent, but there is nothing wrong.

pavelkryukov commented:
The ultimate questions are:
1. Why unified pipeline IPC is not 1?
2. Why non-unified pipeline IPC is not 1?

graudtv commented:
1. Unified pipeline IPC is almost always the same as non-unified one
2. Non-unified pipeline IPC is not 1 because of data hazards

pavelkryukov commented:
> Unified pipeline IPC is almost always the same as non-unified one

What instructions produce data hazards?
As I recall from our lectures, data hazards produce bubbles in 5-cycle pipeline only when load instruction is followed by its consumer. There are no loads, and I would expect everything to be bypassed.

graudtv commented:
> There are no loads, and I would expect everything to be bypassed

Yes, it should be so, I suppose. But I've looked through code and our case is explicitly considered as a reason to stall. The backtrace is the following:
1) addu instruction in decode stage is checked for stall here:
https://github.com/MIPT-ILab/mipt-mips/blob/b17e2944325c9e9ac258a943bc9cf0ffa0bc7aed/simulator/modules/decode/decode.cpp#L125
2) is_stall() returns true, because this condition is met
https://github.com/MIPT-ILab/mipt-mips/blob/b17e2944325c9e9ac258a943bc9cf0ffa0bc7aed/simulator/modules/decode/bypass/data_bypass.h#L50-L52
Values of variables during check are: ```
instruction_latency = Latency( writeback_stage_info.writeback_bandwidth) = 1_lt``` - seems reasonable and ```
writeback_stage_info.operation_latency = 2_lt``` - explicitly set for j instruction, which precedes addu, here:
https://github.com/MIPT-ILab/mipt-mips/blob/b17e2944325c9e9ac258a943bc9cf0ffa0bc7aed/simulator/modules/decode/bypass/data_bypass.h#L145-L148

So it looks like that this logic is written deliberately, but I can't get the intention why it is done so. Do you think it's a bug?

pavelkryukov commented:
> instruction_latency = Latency( writeback_stage_info.writeback_bandwidth) = 1_lt

Should instruction latency be 2_lt for unified pipeline?

> because this condition is met

My guess it prevents completing ADD before/simultaneously with JUMP.

pavelkryukov commented:
> Should instruction latency be 2_lt for unified pipeline?
> My guess it prevent completing ADD before/simultaneously with JUMP.

It might be that we use same function to get data readiness time (2 for cache, 1 for adds) and completion. Probably it should be separated.

graudtv commented:
> Should instruction latency be 2_lt for unified pipeline?

No, because of an extra stage for unified pipeline

> My guess it prevent completing ADD before/simultaneously with JUMP

ADD had writeback stage, it can't complete simultaneously

pavelkryukov commented:
> ADD had writeback stage, it can't complete simultaneously

How do we handle this case?

```
JUMP: F D E B W
ADDU:   F D E W
```

graudtv commented:
You mean that this stall is necessary for non-unified pipeline, but for unified I should eliminate it?
Currently for non-unified it is
```
JUMP: F D E B W
ADDU:   F D - E W
```
and for unified
```
JUMP: F D E B W - -
ADDU:   F D - E - W
```

pavelkryukov commented:
Yes, if each instruction is 5 cycles, there must be no stall, except the load->use case.
